### PR TITLE
Render layers as groups, pt 2

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR continues https://github.com/qgis/QGIS/pull/46154 and adds the necessary backend API pieces to synchronize the layer tree structure with the QgsGroupLayer logic.

No visible UI changes yet!

### CODE REVIEW

File: src/core/layertree/qgslayertreegroup.cpp [hunk 1]
```diff
@@ -306,13 +362,14 @@ void QgsLayerTreeGroup::writeXml( QDomElement &parentElement, const QgsReadWrite
   QDomDocument doc = parentElement.ownerDocument();
   QDomElement elem = doc.createElement( QStringLiteral( "layer-tree-group" ) );
   elem.setAttribute( QStringLiteral( "name" ), mName );
-  elem.setAttribute( QStringLiteral( "expanded" ), mExpanded ? "1" : "0" );
+  elem.setAttribute( QStringLiteral( "expanded" ), mExpanded ? QStringLiteral( "1" ) : QStringLiteral( "0" ) );
```
Comment (nirvn) [inline src/core/layertree/qgslayertreegroup.cpp:365]: QStringLiteral -> QLatin1String?

File: src/core/layertree/qgslayertreegroup.h [hunk 2]
```diff
@@ -139,6 +145,15 @@ class CORE_EXPORT QgsLayerTreeGroup : public QgsLayerTreeNode
      */
     QList<QgsLayerTreeLayer *> findLayers() const;
 
+
+    /**
+     * Returns an ordered list of map layers in the group, ignoring any layers which are child layers of QgsGroupLayers. Searches recursively the whole sub-tree.
```
Comment (nirvn) [inline src/core/layertree/qgslayertreegroup.h:150]: Can you split that long line into two?

**Review by nirvn — APPROVED** [review]

File: src/core/layertree/qgslayertreegroup.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/layertree/qgslayertreegroup.cpp:365]: Nope, QStringLiteral is the correct one here 🤣

File: src/core/layertree/qgslayertreegroup.h (hunk 2 above)
Comment (nyalldawson) [reply src/core/layertree/qgslayertreegroup.h:150]: Sure can 😁

File: src/core/qgsmapsettings.cpp [hunk 3]
```diff
@@ -285,15 +286,42 @@ void QgsMapSettings::setDpiTarget( double dpi )
...
+        expandLayers( groupLayer->childLayers() );
+      }
+      else
+      {
+        result << layer;
+      }
+    }
+  };
+
+  expandLayers( actualLayers );
```
Comment (m-kuhn) [inline src/core/qgsmapsettings.cpp:323]: What is the use for the lambda function here? I see it used only once

**Review by m-kuhn — APPROVED** [review]
Tried to find problems and failed

**Comment by 3nids** [comment]
related to https://github.com/qgis/QGIS-Enhancement-Proposals/issues/235

**Comment by roya0045** [comment]
Could this be related to the slowdown/lagginess when enabling the map hittest?

**Pull request merged**